diff --git a/src/openvpn/forward.c b/src/openvpn/forward.c
index 0443ca0a..dbfd07f2 100644
--- a/src/openvpn/forward.c
+++ b/src/openvpn/forward.c
@@ -926,7 +926,10 @@ read_incoming_link(struct context *c)
 
     status = link_socket_read(c->c2.link_socket,
                               &c->c2.buf,
-                              &c->c2.from);
+			                  &c->c2.from,
+			                  c->options.ce.xormethod,
+			                  c->options.ce.xormask,
+			                  c->options.ce.xormasklen);
 
     if (socket_connection_reset(c->c2.link_socket, status))
     {
@@ -1788,7 +1791,12 @@ process_outgoing_link(struct context *c)
                 socks_preprocess_outgoing_link(c, &to_addr, &size_delta);
 
                 /* Send packet */
-                size = link_socket_write(c->c2.link_socket, &c->c2.to_link, to_addr);
+                size = link_socket_write(c->c2.link_socket,
+                        &c->c2.to_link,
+                        to_addr,
+				        c->options.ce.xormethod,
+				        c->options.ce.xormask,
+				        c->options.ce.xormasklen);
 
                 /* Undo effect of prepend */
                 link_socket_write_post_size_adjust(&size, size_delta, &c->c2.to_link);
